lib/Sema/CompilerRepresentable.cpp

```diff
@@ -0,0 +1,325 @@
+    assert(UnderlyingReason);
+    TC.diagnose(Target, diag::compiler_evaluable_unrepresentable_param,
+                UnrepresentableType, UnderlyingReason->UnrepresentableType);
+    break;
+  case ResultUnrepresentable:
+    assert(UnderlyingReason);
+    TC.diagnose(Target, diag::compiler_evaluable_unrepresentable_result,
+                UnrepresentableType, UnderlyingReason->UnrepresentableType);
+    break;
+  }
+
+  assert(UnderlyingReason);
+  UnderlyingReason->emitDiagnosticNotesImpl(TC, Target);
+}
+
+void UnrepresentableReason::emitDiagnosticNotes(TypeChecker &TC,
+                                                const Decl *Target) const {
+  emitDiagnosticNotesImpl(TC, Target);
+}
+
+void UnrepresentableReason::emitDiagnosticNotes(TypeChecker &TC,
+                                                SourceLoc Target) const {
+  emitDiagnosticNotesImpl(TC, Target);
+}
+
+// Checks if the compile-time evaluator can handle the types that the generic
+// arguments are bound to.
+std::unique_ptr<const UnrepresentableReason>
+CompilerRepresentableChecker::checkGenericArgs(CanBoundGenericType type) {
+  assert(type);
+  for (auto arg : type->getGenericArgs()) {
+    if (auto unrepresentable = check(arg)) {
+      return UnrepresentableReason::getGenericArgUnrepresentable(
+          type, std::move(unrepresentable));
+    }
+  }
+  return NULL;
+}
+
+// Checks if the compile-time evaluator can handle this enum.
+std::unique_ptr<const UnrepresentableReason>
+CompilerRepresentableChecker::checkDeclaredType(const EnumDecl *decl) {
+  assert(decl);
+  for (auto *elementDecl : decl->getAllElements()) {
+    auto *parameterList = elementDecl->getParameterList();
+    if (!parameterList)
+      continue;
+
+    for (auto *paramDecl : *parameterList) {
+      if (auto unrepresentable = check(paramDecl->getType())) {
+        return UnrepresentableReason::getEnumElementUnrepresentable(
+            decl->getDeclaredType(), elementDecl, std::move(unrepresentable));
+      }
+    }
+  }
+  return NULL;
+}
+
+// Checks if the compile-time evaluator can handle this struct.
+std::unique_ptr<const UnrepresentableReason>
+CompilerRepresentableChecker::checkDeclaredType(const StructDecl *decl) {
+  assert(decl);
+  for (auto *fieldDecl : decl->getStoredProperties()) {
+    if (auto unrepresentable = check(fieldDecl->getType())) {
+      return UnrepresentableReason::getFieldUnrepresentable(
+          decl->getDeclaredType(), fieldDecl, std::move(unrepresentable));
+    }
+  }
+  return NULL;
+}
+
+// Checks if the compile-time evaluator can handle the types of the parameters
+// of this function.
+std::unique_ptr<const UnrepresentableReason>
+CompilerRepresentableChecker::checkParams(CanAnyFunctionType type) {
+  assert(type);
+  for (auto param : type->getParams()) {
+    if (auto unrepresentable = check(param.getType())) {
+      return UnrepresentableReason::getParamUnrepresentable(
+          type, std::move(unrepresentable));
+    }
+  }
+  return NULL;
```

Comment:
Use `nullptr` instead of `NULL` (and everywhere else)?

Reply:
+1

Reply:
done